lightyear/src/client/prediction/prespawn.rs

```diff
@@ -374,15 +380,52 @@ impl PreSpawnedPlayerObjectPlugin {
     Component, Serialize, Deserialize, Default, Debug, Copy, Clone, PartialEq, Eq, Reflect,
 )]
 #[component(storage = "SparseSet")]
+/// Added to indicate the client has prespawned the predicted version of this entity.
+///
+/// ```rust,ignore
+/// // Default hashing implementation: (tick + components)
+/// PreSpawnedPlayerObject::default();
+///
+/// // Default hashing implementation with additional user-provided salt:
+/// let client_id: u64 = 12345;
+/// PreSpawnedPlayerObject::default_with_salt(client_id);
+///
+/// // User-provided custom hash
+/// let custom_hash: u64 = compute_hash();
+/// PreSpawnedPlayerObject::new(hash);
+/// ``````
 pub struct PreSpawnedPlayerObject {
     /// The hash that will identify the spawned entity
     /// By default, if the hash is not set, it will be generated from the entity's archetype (list of components) and spawn tick
     /// Otherwise you can manually set it to a value that will be the same on both the client and server
     pub hash: Option<u64>,
+    /// An optional extra value that will be passed to the hasher as part of the default hashing algorithm
+    ///
+    /// Since the default hash uses the tick and components, a useful addition is the client id, to
+    /// distinguish between bullets spawned on the same tick, but by different players.
+    #[serde(skip)]
+    pub user_salt: Option<u64>,
```

Comment:
Do you think I should always add the client_id automatically?

Reply:
probably a good idea, although when non player characters both spawn bullets on the same tick, you'd still want to add a salt with the NPC id too, for example

Reply:
actually the client_id shouldn't be default added because that would make it impossible to predict spawns of global stuff, since the server wouldn't be using your client id.

Reply:
consider the case of a cloud spawning raindrops every 3 seconds. clients could predict those, but if clients include their client id in the hash, it won't match the server. the cloud / raindrops don't belong to a player, so there is no client id to include on the server side.